docs/proposals/error-handling-guideline.md

```diff
@@ -0,0 +1,224 @@
+
+Current behavior and output:
+
+```
+$ oras tag list ghcr.io/oras-project/oras
+Error: unable to add tag for 'list': invalid reference: missing repository
+```
+
+Suggested error message:
+
+```
+$ oras tag list ghcr.io/oras-project/oras
+Error: There is no "list" sub-command for "oras tag" command.
+Usage: oras tag [flags] <name>{:<tag>|@<digest>} <new_tag> [...]
+If you want to tag a manifest in a registry or an OCI image layout, use "oras tag". If you want to list available tags in a repository, use "oras repo show tags" 
+```
+
+#### Example 3: When fetching a manifest if no manifest tag or digest is provided
+
+Current behavior and output:
+
+```
+$ oras manifest fetch --oci-layout /tmp/ginkgo1163328512
+Error: "/tmp/ginkgo1163328512": no tag or digest when expecting <name:tag|name@digest>
+```
+
+Suggested error message:
+
+```
+$ oras manifest fetch --oci-layout /tmp/ginkgo1163328512
+Error: "/tmp/ginkgo1163328512": no tag or digest specified
+Usage: oras manifest fetch [flags] <name>{:<tag>|@<digest>}
+You need to specify an artifact reference in the form of "<name>:<tag>" or "<name>@<digest>". Run "oras manifest fetch -h" for more options and examples 
+```
+
+#### Example 4: push a manifest if no media type flag provided
+
+Current behavior and output:
+
+```
+$ oras manifest push --oci-layout /sample/images:foobar:mediatype
+Error: media type is not recognized. 
+```
+
+Suggested error message:
+
+```
+$ oras manifest push --oci-layout /sample/images:foobar:mediatype
+Error: media type is not recognized. 
+Usage: oras manifest push [flags] <name>[:<tag>[,<tag>][...]|@<digest>] <file>
+You need to specify a valid media type "mediaType" in the manifest JSON or via the "--media-type" flag.
+```
+
+#### Example 5: attach an artifact if the given option is unknown
+
+Current behavior and output:
+
+```
+$ oras attach --artifact-type oras/test localhost:5000/command/images:foobar --distribution-spec v1.0
+Error: unknown distribution specification flag: v1.0
+```
+
+Suggested error message:
+
+```
+$ oras attach --artifact-type oras/test localhost:5000/sample/images:foobar --distribution-spec ???
+Error: unknown distribution specification flag: "v1.0". 
+Available options: v1.1-referrers-api, v1.1-referrers-tag
+```
+
+#### Example 6: when attaching an file, if no file reference is provided
+
+Current behavior and output:
+
+```
+$ oras attach --artifact-type sbom/example localhost:5000/sample/images:foobar
+Error: no blob is provided
+```
+
+Suggested error message:
+
+```
+$ oras attach --artifact-type sbom/example localhost:5000/sample/images:foobar
+Error: failed to attach a file. No file provided in the command.
+Usage: oras attach [flags] --artifact-type=<type> <name>{:<tag>|@<digest>} <file>[:<type>] [...]
+To attach files to an existing artifact, try "oras attach --artifact-type sbom/example localhost:5000/sample/images:foobar sample.json". Run "oras attach -h" for more options and examples
```

Comment:
User need to provide at least a file or an annotation for attaching. The proper error message should be:
```suggestion
$ oras attach --artifact-type sbom/example localhost:5000/sample/images:foobar
Error: no file or annotation provided in the command.
Usage: oras attach [flags] --artifact-type=<type> <name>{:<tag>|@<digest>} <file>[:<type>] [...]
To attach to an existing artifact, please provide files via argument or annotations via flag `-a/--annotation`. Run "oras attach -h" for more options and examples.
```

Reply:
Updated